Tech writing updates for GCP MNIST notebook

sarahmaddox opened the pull request:
Fixes https://github.com/kubeflow/website/issues/1696

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/examples/756)
<!-- Reviewable:end -->

sarahmaddox commented:
/assign @jlewi 
/cc @amygdala

jlewi commented:
Thanks @sarahmaddox Could you clarify the comment about accessing the tensorboard UI? I'd like to make it clear that the cluster role is only needed if you want to programmatically get the URL for the endpoint. However, accessing the endpoint should not depend on the clusterrole.  

The user has to know the endpoint in order to access the notebook and KF so programmatic access is mostly just about being able to print out the link the notebook (its a weird issue with jupyter that its not easy to get it programmatically).

Also it looks like in the command kubectl command given for accessing the mnist UI there is some weird formatting (NAMESPACE is in larger font and in italics?) could you maybe see if you could fix that?

amygdala commented:
Hopefully not muddying the waters, but re: TB, I don't remember getting this error:   "The service account doesn't have sufficient privileges to get the istio-system ingress. You will have to manually enter the Kubeflow endpoint."  or having to run the clusterrole binding command.
I wonder what was different.

sarahmaddox commented on mnist/mnist_gcp.ipynb:
````diff
@@ -1717,37 +1759,39 @@
+    "* If you don't have sufficient priveleges you can grant appropriate permissions by running the following command:\n",
     "  \n",
     "   ```\n",
     "   kubectl create --namespace=istio-system rolebinding --clusterrole=kubeflow-view --serviceaccount=${NAMESPACE}:default-editor ${NAMESPACE}-istio-view\n",
````
@jlewi is this the kubectl command that you mentioned, where the `NAMESPACE` formatting is weird? I haven't actually changed this command. Maybe it's the comment I added below, where `NAMESPACE` is in code formatting? It looks OK in my notebook, so I'm not sure which occurrence of NAMESPACE looks weird in your notebook.

jlewi replied:
Sorry it wasn't due to your changes; it was an existing issue that I just happened to notice.

sarahmaddox commented:
Thanks @jlewi and @amygdala! Here's the exact error that I saw in the notebook:

> ## Access the TensorBoard UI
```
endpoint = k8s_util.get_iap_endpoint() 
if endpoint:    
    vs = yaml.safe_load(tb_virtual_service)
    path= vs["spec"]["http"][0]["match"][0]["uri"]["prefix"]
    tb_endpoint = endpoint + path
    display(HTML(f"TensorBoard UI is at <a href='{tb_endpoint}'>{tb_endpoint}</a>"))
The service account doesn't have sufficient privileges to get the istio-system ingress. You will have to manually enter the Kubeflow endpoint. To make this function work ask someone with cluster priveleges to create an appropriate clusterrolebinding by running a command.
kubectl create --namespace=istio-system rolebinding --clusterrole=kubeflow-view --serviceaccount=${NAMESPACE}:default-editor ${NAMESPACE}-istio-view
``` 

I deployed Kubeflow 1.0 RC4 from the command line on my Mac, not using Cloud Shell.
Deployment date: 14 hours ago. (So, Wed 26 February, Sydney time.)
Config file: https://raw.githubusercontent.com/kubeflow/manifests/master/kfdef/kfctl_gcp_iap.yaml

amygdala commented:
> I deployed Kubeflow 1.0 RC4 from the command line on my Mac, not using Cloud Shell.
> Deployment date: 14 hours ago. (So, Wed 26 February, Sydney time.)
> Config file: https://raw.githubusercontent.com/kubeflow/manifests/master/kfdef/kfctl_gcp_iap.yaml

When I get the chance I might try to repro with my own deployment from yesterday.  The one where it worked for me was an older deployment from 19 days ago.

amygdala commented:
Update: yeah, I hit the same thing with my new deployment.  Seems that recent updates may have changed something.
Jeremy, if they're always going to hit that issue, it might make sense to have them preemptively go use the the cloud shell to define the role binding, before they run that cell. If they're running a codelab, e.g. from a chromebook, they won't have kubectl installed locally. WDYT?

amygdala commented:
That said, I think we should go ahead and pull in Sarah's revisions, as I would like to publish the initial version of the codelab this week.
/lgtm

jlewi commented:
@amygdala @sarahmaddox 

```
The service account doesn't have sufficient privileges to get the istio-system ingress. You will have to manually enter the Kubeflow endpoint. To make this function work ask someone with cluster priveleges to create an appropriate clusterrolebinding by running a command.
```

Is a known issue; see kubeflow/kubeflow#4752

I call it a limitation and not a bug because printing out the URL is just a convenience. The point of the warning that gets displayed is to tell the user what they need to do.

I would prefer not telling the user to run Cloud Shell that breaks the experience.

I think a better solution would be the following.  At the top of the notebook we can add a cell in which the user sets the host name along with markdown that explains how to set it e.g

"Set kf_endpoint to https://<your domain> where your domain will typically be ${KFAPP}.endpoints.${PROJECT}.cloud.goog you can determine this just by looking at the url you are accessing the notebook at" 

```
endpoint = None

if not kf_endpoint:
  logging.info("Warning you must set {endpoint} in order to print out URLs you will access things at.")
```

We can then just remove the markdown telling users to grant the RBAC roles as well as the lines

```
endpoint=k8s_util.get_iap_endpoint()
```

In which case the behavior will be as follows

1. If user sets endpoint it will print out the URLs 
1. If user doesn't set the endpoint they will see a warning and the URLs will not be printed.

jlewi commented:
Before we merge this we should make sure the tests are passing/
Currently the tests are marked as being skipped (they should still run though).

@gabrielwen could you please change the presubmits to be blocking?

jlewi commented:
@amygdala is your codelab blocked on merging this PR? Isn't this PR mostly fit and finish?

amygdala commented:
> @amygdala is your codelab blocked on merging this PR? Isn't this PR mostly fit and finish?

It's not blocked— I'll publish the codelab today without the changes if need be— but it would be nice to get it out soon. 

I like your idea of just having them set the endpoint (optionally).

sarahmaddox commented:
@jlewi and @amygdala I've updated the notebook to set a variable containing the endpoint for the domain, and removed the bits about setting RBAC in order to get the endpoint programmatically. PTAL.

amygdala commented:
Sarah, apologies if I missed something, but is there anywhere you have them actually set that endpoint value?  (`$KUBEFLOW_ENDPOINT`)?  I think we need a cell where they optionally set that, than if it is set, automatically construct the endpoint string. I'm thinking we could pull in your changes, then add that cell in a separate PR if Jeremy is okay with that ordering.

jlewi commented:
I'm fine with this change going in as is and fixing the endpoint in the follow on PR.

Please rebase though to pick up the latest change which runs the test on presubmit. It looks like your branch is based on 
cc93a80420755f4eb09dc505b79c2d1ec77de46d

Which is the commit prior to fixing the test and making it run on presubmit.

sarahmaddox commented:
@amygdala I'd decided to use the `endpoint` variable that the notebook already uses. (I didn't think we needed another variable like  `kf_endpoint` or `kubeflow_endpoint`. From @jlewi 's suggetsed code snippet I couldn't see the difference betweent `endpoint` and `kf_endpoint`.)

There's a Markdown cell asking the user to set the `endpoint` variable, followed by a code cell that they need to adjust to set the variable. It works for getting the TensorBoard endpoint (shown in the screenshot) and for getting the MNIST UI endpoint further down (not in the screenshot).
![image](https://user-images.githubusercontent.com/6138251/75634183-71da4580-5c5f-11ea-828d-f7555f046fa7.png)

It seems to do the trick, but please do let me know if I've got it wrong. :)

sarahmaddox commented:
@jlewi I merged with the examples repo's master. The tests are now running, but are failing. I'm not sure how to fix the errors.

amygdala commented:
> It seems to do the trick, but please do let me know if I got it wrong or should clarify it. :)

That sounds fine (sorry, I just did a string search; I suspected I was missing something). 
/lgtm

jlewi commented:
/test all

Thanks @sarahmaddox 
Test failures look like https://github.com/kubeflow/examples/issues/762

jlewi commented:
Tests should be fixed.
https://k8s-testgrid.appspot.com/sig-big-data#kubeflow-periodic-examples

/lgtm
/approve 
/test all
